oeps/oep-0011.rst

```diff
@@ -0,0 +1,227 @@
+
+Abstract
+========
+
+This document describes the technology best practices for developing
+front end applications and components for Open edX. It covers the techniques
+and approaches that are currently recommended, as well as the technologies
+and libraries that should be used.
+
+Motivation
+==========
+
+The rapid pace of development in front end tooling has created opportunities
+to greatly improve the quality of the Open edX user and developer experience.
+After an assesment of industry best practices, the edX Front End Working Group
+(FedX) has made a number of technology recommendations for libraries, frameworks
+and tooling to modernize the edX front end.
+
+Within this document, the phrase "front end" is used to mean any part
+of the platform that is shown to users. This encompasses views rendered in
+Python on the server, interactive interfaces written using JavaScript, and
+CSS styling.
+
+Specification
+=============
+
+Technology Selection
+~~~~~~~~~~~~~~~~~~~~
+
+1. **Server-side content should be rendered with Django Templates**
+
+   **Rationale**: There are many template languages available for Django,
+   but the simplest option is to use the built-in Django template engine.
+   The Open edX codebase has a mixture of Django and Mako templates, but the
+   former are easier to reason about because they don't support arbitrary
+   code evaluation. See `Reasons for rejecting Mako`_ for a more detailed
+   explanation.
+
+   **Exception**: Mako templates can continue to be used within edx-platform
+   for consistency with the existing code base. This is because the base
+   templates and theming support are all provided via Mako, so it is too
+   much to expect a new feature to be implemented with a different framework.
+   There is much desire to replace Mako within edx-platform so this
+   exception may eventually be removed.
+
+2. **Target the latest standardised JavaScript version (ECMA-262)**
+
+   **Rationale**: ECMA-262 is the specification for JavaScript, and
+   modern front end libraries typically use the version of ECMA-262 released
+   in 2015 (ES2015) or newer. edX JavaScript should use the syntax and
+   features appropriate for the year is is written in, based on ECMA's
+   yearly spec release schedule. Browsers progressively add support for more
+   ECMA-262 features as they are accepted, so it is recommended to use `Babel`_
+   which can transpile ECMA-262-compliant JavaScript into code that all
+   browsers support.
+
+   **Exception**: Much of edX's existing front end code is written conformant
+   to the version of ECMA-262 released in 2009 (ES5). Files written in ES5
+   should be gradually converted to the newer standard as new development in
+   those feature areas requires.
+
+   **Note**: Use of standards-compliant ECMAScript features should be
+   preferred to the use of external libraries such as Underscore or
+   compile-to-JavaScript languages such as CoffeeScript. See
```

Comment:
@ormsbee @bradenmacdonald See what you think of ba63ed5.

(I'm going to move the "use the language's stdlib instead of importing Underscore where possible" into the Developer Guide, as it's common sense.)